### DISCUSSION TITLE
Support nested form data

### DISCUSSION THREAD

**Discussion by aviemet** [inertiajs/inertia#1174] [Ideas]
Since Rails and Laravel both expect form data as nested objects by default, could that be supported by `useForm`?

Currently, the only ways to submit nested data are to either rely on the `transform` method, [which is currently unreliable](https://github.com/inertiajs/inertia/issues/1131), or manually submit the data using `Inertia.post`.

I would like to be able to define a form as such:

```javascript
export default ComponentWithForm = ({ user, person }) => {
  const form = useForm({ user: user, person: person })

  const handleSubmit = e => {
    e.preventDefault()
    form.post('/endpoint')
  }

  return (
    <form onSubmit={ handleSubmit }>
      <input value={ form.user.username } onChange={ e => form.setData('user.username', e.target.value) } />
      <input value={ form.person.name } onChange={ e => form.setData('person.name', e.target.value) } />
    </form>
  )
}
```

Currently, the way I achieve this is as follows:

```javascript
export default ComponentWithForm = ({ user, person }) => {
  const form = useForm({
    username: user.username,
    name: person.name
  })

  const handleSubmit = e => {
    e.preventDefault()

   form.transform(data => ({
      user: {
        username: data.username
      },
      person: {
        name: data.name
      }
    })
    form.post('/endpoint')
  }

  return (
    <form onSubmit={ handleSubmit }>
      <input value={ form.username } onChange={ e => form.setData('username', e.target.value) } />
      <input value={ form.name } onChange={ e => form.setData('name', e.target.value) } />
    </form>
  )
}
```

The issue I have with the current way to achieve this is that I'm breaking down the data to put it into the form, then building it back up in the transform method to submit it back to the server in the original form. I would appreciate the liberty of only using the transform method when I actually need to transform data. Currently, I need to rely on transform every time I submit data since Rails expects form data to be nested in a root property.

Thank you

**Comment by balthild** [comment]
I've implemented a `useInertiaForm` hook that supports dot notation with `dot-prop`, `immer` and some TS type magic: https://gist.github.com/balthild/1f23725059aef8b9231d6c346494b918

Note that it's not a drop-in replacement for `useForm`. It does not support `FormData`. Some of its behaviours are different from `useForm` too. Feel free to improve it and publish it as a support package for `inertia-react` if you have an idea.

**Comment by adamndev** [comment]
+1 for setting nested properties with dot notation 👍🏻

**Comment by sbc640964** [comment]
For this I created my own useForm and inside it the setData function is updated with lodash

```jsx

import {set} from "lodash/fp";
import {useForm as useInertiaForm} from "@inertiajs/inertia-react";

export function useForm(...args){
   const form = useInertiaForm(...args);
   
   function setData(key, value){
     form.setData(
            function (data){
                return set(name, value, data);
            }
        );
   }

  return { ...form, setData}
}
```

**Reply by robzlabz** [reply]
Thanks, my problem solved

**Comment by aviemet** [comment]
[I created a PR](https://github.com/inertiajs/inertia/pull/1338) for this to hopefully start a discussion about whether this should be included. I only changed the React version since I don't know Vue or Svelte.

**Reply by nicklozon** [reply]
I use Svelte `useForm` and it works with nested data.

**Comment by joelpanor** [comment]
Any update?

**Comment by aviemet** [comment]
I created a solution for this and [published it as an npm package](https://github.com/aviemet/useInertiaForm) if it helps anyone. It also has all of the extra things I needed in order to have a full featured form library based on Inertia.

**Reply by thansil-emst** [reply]
Good work man.

**Comment by Aigarsss** [comment]
Big + for this feature. @reinink I think it would be really useful to implement in core.

**Comment by andelwebsolutions** [comment]
Is there any news on this? I would really appreciate this being in the core of inertia.

**Comment by MNhat0308** [comment]
Any update on this ??

**Comment by RiadhKHEDHIRI** [comment]
Up

**Comment by Nelio-Bila** [comment]
Any update on this?

**Comment by arsangamal** [comment]
its easy you can do the following for nested objects

```ts
export default ComponentWithForm = ({ user, person }) => {
  const form = useForm({ user: user, person: person })

  const handleSubmit = e => {
    e.preventDefault()
    form.post('/endpoint')
  }

  return (
    <form onSubmit={ handleSubmit }>
      <input value={ form.user.username } onChange={ e => form.setData('user', {...form.user, username: e.target.value) } />
      <input value={ form.person.name } onChange={ e => form.setData('person', {...form.person, person: e.target.value) } />
    </form>
  )
}
```

**Reply by arsangamal** [reply]
@Nelio-Bila @RiadhKHEDHIRI @MNhat0308 @andelwebsolutions @aviemet Please check my answer

**Reply by Aigarsss** [reply]
@arsangamal This solution would work, but it quickly becomes very inconvenient, especially for complex forms.

**Comment by DGINXREAL** [comment]
Any update?

**Comment by mdthansil** [comment]
This feature is now available by default in Inertia. Update to the latest version of Inertia.
for reference: https://github.com/inertiajs/inertia/pull/2181
